src/node_file.cc

```
@@ -2727,6 +2730,222 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+  CHECK_GE(args.Length(), 1);
+  // packageJSONUrl
+  // /home/user/project/package.json
+  CHECK(args[0]->IsString());
+
+  Environment* env = Environment::GetCurrent(args);
+
+  std::string_view base_path =
+      Utf8Value(env->isolate(), args[0].As<String>()).ToStringView();
+  auto base = ada::parse<ada::url_aggregator>(base_path);
+
+  if (!base) {
+    return args.GetReturnValue().Set(LegacyMainResolveReturnType::kInvalidUrl);
+  }
+
+  ada::url_aggregator* base_pointer = &base.value();
+  ada::url_aggregator* file_path_agg;
+  ada::result<ada::url_aggregator> file_path_url;
+  std::unique_ptr<std::string> initial_file_path;
+  std::string file_path;
+
+  // packageConfig.main
+  // ./index.js
+  // ./lib/index.js
+  if (args[1]->IsString()) {
+    std::string package_config_main =
+        Utf8Value(env->isolate(), args[1].As<String>()).ToString();
+
+    file_path_url = ada::parse<ada::url_aggregator>(
+        std::string("./") + package_config_main, base_pointer);
+
+    if (!file_path_url) {
+      return args.GetReturnValue().Set(LegacyMainResolveReturnType::kInvalidUrl);
+    }
+
+    file_path_agg = &file_path_url.value();
+
+    if (!FileURLToPathImpl(env, file_path_agg, initial_file_path)) return;
+
+    FromNamespacedPath(&(*initial_file_path));
+
+    for (int i = 0; i < legacy_main_extensions_with_main_end; i++) {
+      file_path = *initial_file_path + legacy_main_extensions[i];
+
+      switch(FilePathIsFile(env, file_path)) {
+        case FilePathIsFileReturnType::kIsFile:
+          return args.GetReturnValue().Set(i);
+        case FilePathIsFileReturnType::kIsNotFile:
+          continue;
+        case FilePathIsFileReturnType::kThrowInsufficientPermissions:
+          // the default behavior when do not have permission is to return
+          // and exit the execution of the method as soon as possible
+          // the internal function will throw the exception
+          return;
```

It should never execute in this line. I don't think we need this enum.

---

`FilePathIsFile` now verifies the permission model, I need this case because the user can try to import a library using this old method and that library was protected by the permission model, then it should throw an error.